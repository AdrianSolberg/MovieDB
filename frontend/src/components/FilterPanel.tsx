import { Button } from "@/shadcn/components/ui/button";
import {
    Accordion,
    AccordionContent,
    AccordionItem,
    AccordionTrigger,
} from "../shadcn/components/ui/accordion";
import {
    Sheet,
    SheetClose,
    SheetContent,
    SheetFooter,
    SheetHeader,
    SheetTitle,
    SheetTrigger,
} from "../shadcn/components/ui/sheet";
import { SlidersHorizontal } from "lucide-react";
import { Checkbox } from "@/shadcn/components/ui/checkbox";
import { useEffect, useState } from "react";
import { all_genres, all_languages } from "@/mock/util";
import { Status } from "@/types/movieTypes";

const FilterPanel = () => {
    const [filters, setFilters] = useState<{ [key: string]: string[] }>({});

    const all_filters: { [key: string]: string[] } = {
        Genre: all_genres.map((genre) => genre.name),
        Rating: ["5", "4", "3", "2", "1"],
        "Release Year": [
            "2020s",
            "2010s",
            "2000s",
            "1990s",
            "1980s",
            "1970s",
            "1960s",
            "1950s",
            "1940s",
            "1930s",
            "1920s",
            "1910s",
            "1900s",
            "1890s",
            "1880s",
            "1870s",
        ],
        Language: all_languages.map((language) => language.name),
        Status: Object.values(Status) as string[],
        Runtime: ["Less than 1 hour", "1 - 2 hours", "More than 2 hours"],
    };

    useEffect(() => {
        const storedFilters = sessionStorage.getItem("filters");
        if (storedFilters) {
            setFilters(JSON.parse(storedFilters));
        }
    }, []);

    const updateFilters = (category: string, filter: string) => {
        let newFilters: string[] = [...(filters[category] || [])];
        if (newFilters.includes(filter)) {
            newFilters = newFilters.filter((e) => e != filter);
        } else {
            newFilters.push(filter);
        }
        const updatedFilters = { ...filters, [category]: newFilters };
        sessionStorage.setItem("filters", JSON.stringify(updatedFilters));
        setFilters(updatedFilters);
    };

    return (
        <Sheet>
            <SheetTrigger>
                <Button variant="outline" size="sm" className="h-7 gap-1">
                    <SlidersHorizontal className="h-3.5 w-3.5" />
                    <span className="sr-only sm:not-sr-only sm:whitespace-nowrap">
                        Filters
                    </span>
                </Button>
            </SheetTrigger>
            <SheetContent side="left" className="w-full overflow-y-auto">
                <SheetHeader className="mb-5">
                    <SheetTitle>Filters</SheetTitle>
                </SheetHeader>
                <Accordion type="single" collapsible className="w-full">
                    {Object.entries(all_filters).map(
                        ([category, filter_list]) => (
                            <AccordionItem key={category} value={`${category} item`}>
                                <AccordionTrigger>
                                    {category}
                                    {filters[category]?.length
                                        ? ` (${filters[category].length})`
                                        : ""}
                                </AccordionTrigger>
                                <AccordionContent>
                                    <ul>
                                        {filter_list.map((filter) => (
                                            <li
                                                key={filter}
                                                className="flex items-center space-x-2 mb-2"
                                            >
                                                <Checkbox
                                                    id={filter}
                                                    checked={filters[
                                                        category
                                                    ]?.includes(filter)}
                                                    onCheckedChange={() =>
                                                        updateFilters(
                                                            category,
                                                            filter
                                                        )
                                                    }
                                                />
                                                <label
                                                    htmlFor={filter}
                                                    className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                                                >
                                                    {filter}
                                                </label>
                                            </li>
                                        ))}
                                    </ul>
                                </AccordionContent>
                            </AccordionItem>
                        )
                    )}
                </Accordion>
                <SheetFooter className="mt-5">
                    <SheetClose asChild>
                        <Button type="submit">Apply</Button>
                    </SheetClose>
                </SheetFooter>
            </SheetContent>
        </Sheet>
    );
};

export default FilterPanel;
