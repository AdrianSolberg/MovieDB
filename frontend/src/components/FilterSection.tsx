import { Filter, FiltersInput } from "@/types/__generated__/types";
import {
    AccordionContent,
    AccordionItem,
    AccordionTrigger,
} from "../shadcn/components/ui/accordion";
import { Checkbox } from "../shadcn/components/ui/checkbox";
import { formatNumber } from "@/utils/formatUtil";

interface FilterSectionInterface {
    category: keyof FiltersInput;
    all_filters: Filter[];
    applied_filters: string[];
    updateFilters: (category: keyof FiltersInput, filter: string) => void;
    loading: boolean;
}

const FilterSection: React.FC<FilterSectionInterface> = ({
    category,
    all_filters,
    applied_filters,
    updateFilters,
    loading,
}) => {
    return (
        <AccordionItem value={`${category} item`}>
            <AccordionTrigger>
                <p>
                    {category}
                    {applied_filters.length > 0 && (
                        <span className="opacity-60">
                            {" "}
                            ({applied_filters.length} applied)
                        </span>
                    )}
                </p>
            </AccordionTrigger>
            <AccordionContent>
                <ul>
                    {all_filters.map(
                        (filter) =>
                            filter.hits > 0 && (
                                <li
                                    key={filter.name}
                                    className="mb-2 flex items-center space-x-2"
                                >
                                    <Checkbox
                                        id={filter.name}
                                        checked={applied_filters.includes(
                                            filter.name
                                        )}
                                        disabled={loading}
                                        onCheckedChange={() =>
                                            updateFilters(category, filter.name)
                                        }
                                    />
                                    <label
                                        htmlFor={filter.name}
                                        className="text-sm font-medium leading-none hover:cursor-pointer"
                                    >
                                        {filter.name}{" "}
                                        <span className="opacity-60">
                                            ({formatNumber(filter.hits)})
                                        </span>
                                    </label>
                                </li>
                            )
                    )}
                </ul>
            </AccordionContent>
        </AccordionItem>
    );
};

export default FilterSection;
