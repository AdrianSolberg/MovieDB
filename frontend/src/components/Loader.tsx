import { cn } from "@/shadcn/lib/utils";
import React from "react";

type LoaderProps = {
    color?: "stroke-primary" | "stroke-secondary";
    size?: "sm" | "md" | "lg";
} & React.HTMLAttributes<HTMLDivElement>;

const Loader = React.forwardRef<
    HTMLDivElement,
    React.PropsWithChildren<LoaderProps>
>(
    (
        {
            className,
            children,
            color = "stroke-primary",
            size = "md",
            ...props
        },
        ref
    ) => {
        const sizeClass =
            size === "sm"
                ? "h-6 w-6"
                : size === "lg"
                  ? "h-16 w-16"
                  : "h-12 w-12";

        return (
            <div
                ref={ref}
                className={cn(
                    "m-2 flex flex-col items-center justify-center gap-2",
                    className
                )}
                role="status"
                aria-live="polite"
                {...props}
            >
                <svg
                    viewBox="0 0 800 800"
                    xmlns="http://www.w3.org/2000/svg"
                    className={cn(
                        "animate-spin stroke-primary",
                        sizeClass,
                        color
                    )}
                >
                    <circle
                        cx="400"
                        cy="400"
                        fill="none"
                        r="350"
                        strokeWidth="64"
                        strokeDasharray="1648,550"
                        strokeLinecap="round"
                    />
                </svg>
                {children ? (
                    children
                ) : (
                    <span className="sr-only">Loading...</span>
                )}
            </div>
        );
    }
);

export default Loader;
