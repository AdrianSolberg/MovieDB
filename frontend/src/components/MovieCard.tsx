import {
    Card,
    CardContent,
    CardDescription,
    CardHeader,
} from "@/shadcn/components/ui/card";
import Ratings from "@/shadcn/components/ui/rating";
import { Movie } from "@/types/movieTypes";
import { Link } from "react-router-dom";

const MovieCard = ({ movie }: { movie: Movie }) => {
    return (
        <Link to={`/movie/${movie._id}`}>
            <Card className="shadow-lg hover:shadow-slate-600 hover:cursor-pointer m-0">
                <CardHeader className="p-2">
                    <img
                        className="w-full"
                        src={
                            "https://image.tmdb.org/t/p/w500" +
                            movie.poster_path
                        }
                        alt={movie.title}
                    />
                </CardHeader>
                <CardContent className="p-2">
                    <CardDescription className="text-center">
                        {movie.release_date.getFullYear()} â€¢{" "}
                        {Math.floor(movie.runtime / 60)}h {movie.runtime % 60}m
                    </CardDescription>
                    <Ratings
                        className="flex justify-around"
                        value={movie.vote_average / 2}
                        variant="yellow"
                        totalstars={5}
                    />
                </CardContent>
            </Card>
        </Link>
    );
};

export default MovieCard;
