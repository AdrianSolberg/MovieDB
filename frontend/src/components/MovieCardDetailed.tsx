import React from "react";
import {
    Card,
    CardContent,
    CardHeader,
    CardTitle,
} from "../shadcn/components/ui/card";
import { Button } from "../shadcn/components/ui/button";
import Ratings from "../shadcn/components/ui/rating";
import { getImageUrl, ImageType } from "@/utils/imageUrl/imageUrl";
import { Movie } from "@/types/__generated__/types";

interface MovieCardDetailedProps {
    movie: Movie;
}

const formatCurrency = (value: number) => {
    return value.toLocaleString("en-US", {
        style: "currency",
        currency: "USD",
        maximumFractionDigits: 0,
    });
};

const getMovieDetailsList = (movie: Movie) => [
    {
        label: "Status",
        value: movie.status,
    },
    {
        label: "Budget",
        value: movie.budget ? `${formatCurrency(movie.budget)}` : "N/A",
    },
    {
        label: "Revenue",
        value: movie.revenue ? `${formatCurrency(movie.revenue)}` : "N/A",
    },
    {
        label: "Production Companies",
        value: movie.production_companies.length
            ? movie.production_companies.join(", ")
            : "N/A",
    },
    {
        label: "Production Countries",
        value: movie.production_countries.length
            ? movie.production_countries.join(", ")
            : "N/A",
    },
    {
        label: "Spoken Languages",
        value: movie.spoken_languages.length
            ? movie.spoken_languages.join(", ")
            : "N/A",
    },
];

const MovieCardDetailed: React.FC<MovieCardDetailedProps> = ({ movie }) => {
    const releaseYear = movie.release_date.getFullYear();
    const runtimeHours = Math.floor(movie.runtime / 60);
    const runtimeMinutes = movie.runtime % 60;

    return (
        <Card
            className="relative m-4 overflow-hidden bg-cover bg-center shadow-lg"
            style={{
                backgroundImage: `url(${getImageUrl(ImageType.BACKDROP, movie.backdrop_path)})`,
            }}
            data-testid="detailedMovieCard"
        >
            {movie.backdrop_path && (
                <div className="absolute inset-0 bg-black opacity-85" />
            )}
            <CardContent className="relative h-full p-6 text-white">
                <CardHeader className="mb-4 p-0">
                    <CardTitle className="text-5xl font-bold">
                        {movie.title}
                    </CardTitle>
                    <p className="text-sm">
                        {[
                            releaseYear,
                            `${runtimeHours}h ${runtimeMinutes}m`,
                            movie.genres.length
                                ? movie.genres.join(", ")
                                : null,
                        ]
                            .filter(Boolean)
                            .join(" â€¢ ")}
                    </p>
                    <Ratings
                        value={movie.vote_average / 2}
                        size={24}
                        variant="yellow"
                        totalstars={5}
                        title={String((movie.vote_average / 2).toFixed(2))}
                    />
                </CardHeader>
                <section className="flex h-full flex-col md:flex-row">
                    <figure className="mb-6 w-full max-w-72 flex-shrink-0 md:mb-0 md:mr-6">
                        <img
                            src={getImageUrl(
                                ImageType.POSTER,
                                movie.poster_path,
                                "w500"
                            )}
                            className="rounded-lg"
                            alt={`Poster of ${movie.title}`}
                            title={movie.title}
                        />
                    </figure>
                    <section className="flex-grow">
                        {movie.tagline && (
                            <p className="mb-2 text-xl italic text-gray-300">
                                {movie.tagline}
                            </p>
                        )}
                        <p className="mb-4 text-lg">{movie.overview}</p>
                        <ul className="mb-6 flex flex-col gap-2">
                            {getMovieDetailsList(movie).map((item, index) => (
                                <li key={index}>
                                    <strong>{item.label}: </strong>
                                    {item.value}
                                </li>
                            ))}
                        </ul>
                        {movie.homepage && (
                            <Button asChild>
                                <a
                                    href={movie.homepage}
                                    target="_blank"
                                    rel="noopener noreferrer"
                                >
                                    Visit Homepage
                                </a>
                            </Button>
                        )}
                    </section>
                </section>
            </CardContent>
        </Card>
    );
};

export default MovieCardDetailed;
