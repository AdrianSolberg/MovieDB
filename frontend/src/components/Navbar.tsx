import { Link } from "react-router-dom";
import logo from "../assets/logo.png";
import NavbarOverlay from "./NavbarOverlay";
import UserDropdown from "./UserDropdown";
import { useReactiveVar } from "@apollo/client";
import { usernameVar } from "@/utils/cache";

export type NavbarProps = { name: string; href: string }[];

const Navbar = () => {
    const username = useReactiveVar(usernameVar);
    const links: NavbarProps = [
        { name: "Activity", href: "/activity" },
        ...(username != "Guest"
            ? [{ name: "My Reviews", href: "/myReviews" }]
            : []),
    ];

    return (
        <header className="flex w-full items-center justify-between bg-gradient-to-r from-muted-foreground to-primary p-4">
            <nav className="flex items-center gap-10 text-background">
                <Link to="/" className="flex flex-row gap-3 text-2xl font-bold">
                    <img src={logo} alt="logo" className="h-8" />
                    MovieDB
                </Link>
                <ul className="hidden gap-4 text-lg sm:flex">
                    {links.map((link) => (
                        <li key={link.name}>
                            <Link to={link.href} className="hover:underline">
                                {link.name}
                            </Link>
                        </li>
                    ))}
                </ul>
            </nav>
            <div className="flex items-center">
                <UserDropdown />
                {links.length > 0 && (
                    <div className="ml-2">
                        <NavbarOverlay links={links} />
                    </div>
                )}
            </div>
        </header>
    );
};

export default Navbar;
