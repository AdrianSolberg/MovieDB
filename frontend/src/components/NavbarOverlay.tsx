import { Button } from "@/shadcn/components/ui/button";
import {
    Sheet,
    SheetClose,
    SheetContent,
    SheetDescription,
    SheetTitle,
    SheetTrigger,
} from "@/shadcn/components/ui/sheet";
import { Menu } from "lucide-react";
import { Link } from "react-router-dom";
import { NavbarProps } from "./Navbar";

const NavbarOverlay = ({ links }: { links: NavbarProps }) => {
    // This component will show a hamburger menu on small screens
    // Clicking the menu will show an overlay with the navigation links
    return (
        <Sheet>
            <SheetTrigger asChild>
                <Button variant="outline" size="icon" className="sm:hidden">
                    <Menu />
                    <span className="sr-only">Toggle navigation menu</span>
                </Button>
            </SheetTrigger>
            <SheetTitle />
            <SheetDescription />
            <SheetContent side="right">
                <nav className="p-4 shadow-sm">
                    <ul className="flex flex-col space-y-4">
                        {links.map((link) => (
                            <li key={link.name}>
                                {/* Clicking a link should close the overlay */}
                                <SheetClose asChild>
                                    <Link
                                        to={link.href}
                                        className="text-foreground transition-colors hover:text-primary"
                                    >
                                        {link.name}
                                    </Link>
                                </SheetClose>
                                <hr />
                            </li>
                        ))}
                    </ul>
                </nav>
            </SheetContent>
        </Sheet>
    );
};

export default NavbarOverlay;
