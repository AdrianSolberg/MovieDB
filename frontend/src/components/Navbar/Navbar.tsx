import NavbarOverlay from "./NavbarOverlay";
import UserDropdown from "./UserDropdown";

export type NavbarProps = { name: string; href: string }[];

const Navbar = () => {
    const links: NavbarProps = [
        { name: "Movies", href: "/movies" },
        { name: "Feed", href: "/feed" },
    ];

    return (
        <header className="sticky top-0 left-0 z-50 bg-muted shadow-sm w-dvw flex items-center p-4">
            <nav className="flex-1 flex justify-center items-center gap-10">
                <a href="/" className="text-2xl font-bold text-primary">
                    Logo
                </a>
                <ul className="hidden sm:flex gap-4 text-lg">
                    {links.map((link) => (
                        <li key={link.name}>
                            <a
                                href={link.href}
                                className="hover:text-primary text-foreground transistion-colors"
                            >
                                {link.name}
                            </a>
                        </li>
                    ))}
                </ul>
            </nav>
            <div className="absolute right-4 flex items-center gap-2">
                <UserDropdown />
                <NavbarOverlay links={links} />
            </div>
        </header>
    );
};

export default Navbar;
