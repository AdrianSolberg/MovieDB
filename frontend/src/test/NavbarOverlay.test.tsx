import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { NavbarProps } from "@/components/Navbar";
import NavbarOverlay from "@/components/NavbarOverlay";

describe("NavbarOverlay", () => {
    it("successfully renders NavbarOverlay with links", async () => {
        const mockLinks: NavbarProps = [
            { name: "Home", href: "/" },
            { name: "About", href: "/about" },
            { name: "Contact", href: "/contact" },
        ];
        render(
            <MemoryRouter>
                <NavbarOverlay links={mockLinks} />
            </MemoryRouter>
        );

        expect(screen.getByText("Toggle navigation menu")).toBeInTheDocument();

        // Opens sheeet
        await userEvent.click(screen.getByText("Toggle navigation menu"));
        expect(screen.getByText("Home")).toBeInTheDocument();
        expect(screen.getByText("About")).toBeInTheDocument();
        expect(screen.getByText("Contact")).toBeInTheDocument();

        mockLinks.forEach((link) => {
            expect(screen.getByText(link.name)).toBeInTheDocument();
            expect(
                screen.getByRole("link", { name: link.name })
            ).toHaveAttribute("href", link.href);
        });

        // Closes sheet
        expect(screen.getByText("Close")).toBeInTheDocument();
        await userEvent.click(screen.getByText("Close"));
        expect(screen.queryByText("Home")).toBeNull();
        expect(screen.queryByText("About")).toBeNull();
        expect(screen.queryByText("Contact")).toBeNull();
    });

    // Does not happen in practice, but nice to check the component can still handle it
    it("successfully renders NavbarOverlay without links", async () => {
        const mockLinks: NavbarProps = [];
        render(
            <MemoryRouter>
                <NavbarOverlay links={mockLinks} />
            </MemoryRouter>
        );
        expect(screen.getByText("Toggle navigation menu")).toBeInTheDocument();
        await userEvent.click(screen.getByText("Toggle navigation menu"));
        expect(screen.getByText("Close")).toBeInTheDocument();
        await userEvent.click(screen.getByText("Close"));
        expect(screen.queryByText("Close")).toBeNull();
    });
});
