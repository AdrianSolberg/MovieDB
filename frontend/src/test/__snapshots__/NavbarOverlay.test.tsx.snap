// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavbarOverlay > matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10 sm:hidden"
    data-state="closed"
    type="button"
  >
    <svg
      class="lucide lucide-menu"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="4"
        x2="20"
        y1="12"
        y2="12"
      />
      <line
        x1="4"
        x2="20"
        y1="6"
        y2="6"
      />
      <line
        x1="4"
        x2="20"
        y1="18"
        y2="18"
      />
    </svg>
    <span
      class="sr-only"
    >
      Toggle navigation menu
    </span>
  </button>
  <h2
    class="text-lg font-semibold text-foreground"
    id="radix-:r1:"
  />
  <p
    class="text-sm text-muted-foreground"
    id="radix-:r2:"
  />
</DocumentFragment>
`;
